### PR TITLE
 [Reply] Implement email model

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
This pull request pulls in the email model and store from https://github.com/flschweiger/reply :
* Add provider as a dependency
* Add email model
* Add store for emails
* Link email model to MailViewPage 
* Link email model to MailPreviewCard using provider
* Link email store to InboxPage using provider
* Cleanup layout of MailPreviewCard
* Cleanup layout of MailViewPage
* Readability fixes

reply-demo2

### CODE REVIEW

File: lib/studies/reply/mail_card_preview.dart [hunk 1]
```diff
@@ -1,14 +1,23 @@
...
 const _assetsPackage = 'flutter_gallery_assets';
 const _iconAssetLocation = 'reply/icons';
 
 class MailPreviewCard extends StatelessWidget {
-  const MailPreviewCard({Key key}) : super(key: key);
+  const MailPreviewCard({
+    Key key,
+    @required this.id,
+    @required this.email,
+  }) : super(key: key);
```
Comment (JoseAlba) [inline lib/studies/reply/mail_card_preview.dart]: nit. assert that id and email is not null.

File: lib/studies/reply/model/email_model.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  const Email(
+      {this.sender,
+      this.time,
+      this.subject,
+      this.message,
+      this.avatar,
+      this.recipients,
+      this.hasAttachment,
+      this.containsPictures,
+      this.isRead});
```
Comment (JoseAlba) [inline lib/studies/reply/model/email_model.dart]: add final comma

**Review by JoseAlba — APPROVED** [review]
LGTM

File: lib/studies/reply/inbox.dart [hunk 3]
```diff
@@ -2,41 +2,51 @@ import 'package:flutter/material.dart';
 import 'package:gallery/layout/adaptive.dart';
 import 'package:gallery/studies/reply/mail_card_preview.dart';
 
+import 'package:provider/provider.dart';
+import 'package:gallery/studies/reply/model/email_store.dart';
```
Comment (guidezpl) [inline lib/studies/reply/inbox.dart:5]: nit: keep import order consistent

File: lib/studies/reply/inbox.dart [hunk 4]
```diff
@@ -2,41 +2,51 @@ import 'package:flutter/material.dart';
...
-                  child: ListView.separated(
-                    padding: EdgeInsetsDirectional.only(
-                      start: horizontalPadding,
-                      end: horizontalPadding,
-                      top: isDesktop ? 28 : 6,
+    return Consumer<EmailStore>(
+      builder: (context, model, child) {
+        return Navigator(
+          onGenerateRoute: (settings) {
+            return MaterialPageRoute<void>(builder: (context) {
```
Comment (guidezpl) [inline lib/studies/reply/inbox.dart]: nit: formatting. More broadly, I think having 3 builders here is a bit much. Consider splitting this into an inbox navigator widget (or moving the navigator logic to InboxPage's parent) and the UI inbox, which would be Consumer  → SafeArea  → Row...

File: lib/studies/reply/mail_card_preview.dart [hunk 5]
```diff
@@ -26,72 +38,95 @@ class MailPreviewCard extends StatelessWidget {
...
+                                  ),
+                                ),
+                              ),
+                            ],
                           ),
-                          SizedBox(width: 20),
-                          Icon(Icons.more_vert, color: ReplyColors.blue700),
-                        ],
+                        ),
+                      ],
```
Comment (guidezpl) [inline lib/studies/reply/mail_card_preview.dart]: Would benefit in readability for the avatar to be its own StatelessWidget

File: lib/studies/reply/mail_card_preview.dart [hunk 6]
```diff
@@ -26,72 +38,95 @@ class MailPreviewCard extends StatelessWidget {
...
-                      child: ClipOval(
-                        child: Image.asset(
-                          'reply/avatars/avatar_0.jpg',
-                          package: _assetsPackage,
-                          height: 36,
-                          width: 36,
+                      child: Text(
+                        email.message.substring(
+                          0,
+                          email.message.indexOf(RegExp(r'[\n\n]|!|[.]'), 0) + 1,
```
Comment (guidezpl) [inline lib/studies/reply/mail_card_preview.dart]: this is like voodoo to me, what's actually going on? :P

File: lib/studies/reply/mail_card_preview.dart [hunk 7]
```diff
@@ -26,72 +38,95 @@ class MailPreviewCard extends StatelessWidget {
...
+                                    ImageIcon(
+                                      AssetImage(
+                                        '$_iconAssetLocation/twotone_delete.png',
+                                        package: _assetsPackage,
+                                      ),
+                                      color: ReplyColors.blue600,
+                                    ),
+                                    SizedBox(width: 20),
+                                    Icon(Icons.more_vert,
+                                        color: ReplyColors.blue700),
```
Comment (guidezpl) [inline lib/studies/reply/mail_card_preview.dart]: nit: trailing comma

File: pubspec.yaml [hunk 8]
```diff
@@ -29,6 +29,7 @@ dependencies:
   google_fonts: ^1.0.0
   flutter_staggered_grid_view: ^0.3.0
   animations: ^1.1.2
+  provider: ^4.3.2
```
Comment (guidezpl) [inline pubspec.yaml:32]: please also commit pubspec.lock

File: lib/studies/reply/mail_view_page.dart [hunk 9]
```diff
@@ -1,8 +1,15 @@
 import 'package:flutter/material.dart';
 import 'package:gallery/studies/reply/colors.dart';
 
+import 'package:gallery/studies/reply/model/email_model.dart';
+
 class MailViewPage extends StatelessWidget {
-  const MailViewPage({Key key}) : super(key: key);
+  const MailViewPage({Key key, @required this.id, @required this.email})
+      : super(key: key);
```
Comment (guidezpl) [inline lib/studies/reply/mail_view_page.dart]: nit: add non-null asserts

**Review by guidezpl — CHANGES_REQUESTED** [review]

File: lib/studies/reply/inbox.dart (hunk 3 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/inbox.dart:5]: Will do!

File: lib/studies/reply/mail_card_preview.dart (hunk 6 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/mail_card_preview.dart]: Haha it's searching for the first occurrence of a `.` or `!` or `\n\n' and returns a substring from index 0 to the index of that first occurrence. This lets me take part of the email message and show it on the preview card. There is probably a cleaner way to do this so i'll revisit this.

File: pubspec.yaml (hunk 8 above)
Comment (Renzo-Olivares) [reply pubspec.yaml:32]: Will do!

File: lib/studies/reply/mail_card_preview.dart (hunk 5 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/mail_card_preview.dart]: Will do

File: lib/studies/reply/inbox.dart (hunk 4 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/inbox.dart]: I think you're right. I'll update this implementation thanks!

File: lib/studies/reply/mail_card_preview.dart (hunk 6 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/mail_card_preview.dart]: Updated the implementation!

File: lib/studies/reply/adaptive_nav.dart [hunk 10]
```diff
@@ -521,7 +520,15 @@ class _MobileNavState extends State<_MobileNav> with TickerProviderStateMixin {
       overflow: Overflow.visible,
       key: _bottomDrawerKey,
       children: [
-        mainLayer,
+        Navigator(
```
Comment (rami-a) [inline lib/studies/reply/adaptive_nav.dart:523]: Can you remind me again why this needs a new navigator?

File: lib/studies/reply/mail_view_page.dart [hunk 11]
```diff
@@ -82,3 +40,94 @@ class MailViewPage extends StatelessWidget {
...
+        assert(sender != null),
+        assert(recipients != null),
+        assert(time != null),
+        assert(avatar != null);
+
+  final String subject;
+  final String sender;
+  final String recipients;
+  final String time;
+  final String avatar;
```
Comment (rami-a) [inline lib/studies/reply/mail_view_page.dart]: Maybe just pass the `Email` object instead of 5 separate params?

**Review by rami-a — APPROVED** [review]
LGTM

File: lib/studies/reply/adaptive_nav.dart (hunk 10 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/adaptive_nav.dart:523]: To keep the bottom app bar or navigation rail/drawer persistent when navigating to the MailViewPage. Without the navigator the incoming page will take up the whole scaffold covering the bottom app bar or navigation rail.

File: lib/studies/reply/mail_view_page.dart (hunk 11 above)
Comment (Renzo-Olivares) [reply lib/studies/reply/mail_view_page.dart]: That makes sense, wasn't sure what was preferred.

**Pull request merged**